File: bridge/bridge.pl
#!/usr/bin/perl
# Claude Firefox Bridge - local command mailbox.
# The Firefox extension short-polls GET /poll and POSTs results to /result.
# Claude (via curl) POSTs to /command and reads GET /result/<id>.
# No external deps: IO::Socket::INET, JSON::PP, MIME::Base64 are all core.

use strict;
use warnings;
use IO::Socket::INET;
use JSON::PP;
use MIME::Base64 qw(decode_base64);
use File::Basename;
use Cwd qw(abs_path);

my $PORT = $ARGV[0] || 8765;

# Resolve project dir from this script's location so screenshots land in ../shots
my $scriptdir = dirname(abs_path($0));
my $projdir   = dirname($scriptdir);
my $shots_cyg = "$projdir/shots";
mkdir $shots_cyg unless -d $shots_cyg;

# Windows form of the shots dir (for paths handed back to the Read tool)
my $shots_win = $shots_cyg;
$shots_win =~ s{^/([a-zA-Z])/}{uc($1).":\\"}e;   # /c/ -> C:\
$shots_win =~ s{/}{\\}g;

my $json = JSON::PP->new->utf8->canonical;

my @queue;            # commands waiting for the extension
my %results;          # id => result (Perl structure)
my $counter = 0;

my $srv = IO::Socket::INET->new(
    LocalAddr => '127.0.0.1',
    LocalPort => $PORT,
    Listen    => 32,
    ReuseAddr => 1,
    Proto     => 'tcp',
) or die "Cannot bind 127.0.0.1:$PORT : $!\n";

print "Claude Firefox Bridge listening on http://localhost:$PORT\n";
print "Screenshots -> $shots_win\n";

sub send_response {
    my ($cli, $code, $ctype, $body) = @_;
    $body = '' unless defined $body;
    my $len = length($body);
    my $msg = "HTTP/1.1 $code\r\n"
        . "Content-Type: $ctype\r\n"
        . "Content-Length: $len\r\n"
        . "Access-Control-Allow-Origin: *\r\n"
        . "Access-Control-Allow-Methods: GET, POST, OPTIONS\r\n"
        . "Access-Control-Allow-Headers: content-type\r\n"
        . "Connection: close\r\n"
        . "\r\n"
        . $body;
    print $cli $msg;
}

sub send_json {
    my ($cli, $code, $data) = @_;
    send_response($cli, $code, 'application/json', $json->encode($data));
}

while (my $cli = $srv->accept()) {
    $cli->autoflush(1);

    # --- read request line + headers ---
    my $reqline = <$cli>;
    unless (defined $reqline) { close $cli; next; }
    $reqline =~ s/\r?\n$//;
    my ($method, $path) = $reqline =~ m{^(\S+)\s+(\S+)};
    $method ||= ''; $path ||= '';

    my %h;
    while (my $line = <$cli>) {
        $line =~ s/\r?\n$//;
        last if $line eq '';
        if ($line =~ /^([\w-]+):\s*(.*)$/) { $h{lc $1} = $2; }
    }
    my $body = '';
    if (my $len = $h{'content-length'}) {
        read($cli, $body, $len) if $len > 0;
    }

    # --- routes ---
    if ($method eq 'OPTIONS') {
        send_response($cli, '204 No Content', 'text/plain', '');
        close $cli; next;
    }

    if ($method eq 'GET' && $path eq '/health') {
        send_json($cli, '200 OK', {
            ok => JSON::PP::true, queued => scalar(@queue),
            pending_results => scalar(keys %results), port => $PORT,
        });
    }
    elsif ($method eq 'GET' && $path eq '/poll') {
        if (@queue) {
            my $cmd = shift @queue;
            send_json($cli, '200 OK', $cmd);
        } else {
            send_response($cli, '200 OK', 'application/json', '{}');
        }
    }
    elsif ($method eq 'POST' && $path eq '/command') {
        my $data = eval { $json->decode($body) } || {};
        $counter++;
        my $id = $counter;
        push @queue, { id => $id, action => ($data->{action} // ''), params => ($data->{params} // {}) };
        send_json($cli, '200 OK', { id => $id });
    }
    elsif ($method eq 'POST' && $path eq '/result') {
        my $data = eval { $json->decode($body) };
        if ($data && defined $data->{id}) {
            my $result = $data->{result};
            # If the extension returned a screenshot, write it to disk and
            # hand back a file path instead of a huge base64 blob.
            if (ref($result) eq 'HASH' && defined $result->{png_base64}) {
                my $id = $data->{id};
                my $file_cyg = "$shots_cyg/shot-$id.png";
                my $file_win = "$shots_win\\shot-$id.png";
                if (open(my $fh, '>:raw', $file_cyg)) {
                    print $fh decode_base64($result->{png_base64});
                    close $fh;
                    delete $result->{png_base64};
                    $result->{png_path} = $file_win;
                } else {
                    $result->{error} = "could not write screenshot: $!";
                    delete $result->{png_base64};
                }
            }
            $results{ $data->{id} } = $result;
            send_json($cli, '200 OK', { ok => JSON::PP::true });
        } else {
            send_json($cli, '400 Bad Request', { error => 'missing id' });
        }
    }
    elsif ($method eq 'GET' && $path =~ m{^/result/(\d+)$}) {
        my $id = $1;
        if (exists $results{$id}) {
            my $r = delete $results{$id};
            send_json($cli, '200 OK', { done => JSON::PP::true, result => $r });
        } else {
            send_json($cli, '200 OK', { done => JSON::PP::false });
        }
    }
    elsif ($method eq 'GET' && ($path eq '/' || $path eq '/favicon.ico')) {
        send_response($cli, '200 OK', 'text/plain', "Claude Firefox Bridge running on port $PORT\n");
    }
    else {
        send_json($cli, '404 Not Found', { error => 'no such route', method => $method, path => $path });
    }

    close $cli;
}

File: server/bridge.js
'use strict';
/**
 * Claude Firefox Bridge - local command mailbox.
 *
 * Bentaco the Destroyer's Firefox connector.
 *
 * The Firefox extension short-polls GET /poll and POSTs results to /result.
 * The MCP server POSTs to /command and reads GET /result/<id>.
 *
 * Node, standard library only - no npm install, no Perl, no Git Bash, no Python.
 * Node ships with Claude, so this works on a fresh machine with nothing installed.
 *
 * Every hard-won behaviour from the original is preserved deliberately:
 *   * GENERATION FENCING - a reloaded background page bumps the generation and any
 *     older (zombie) poll loop gets 410 and halts. Without it two consumers drain the
 *     same single-consumer queue and commands vanish.
 *   * AT-LEAST-ONCE DELIVERY - /poll moves a command in-flight, /ack clears it, and
 *     anything un-acked for 35s is re-queued. Do NOT eagerly requeue on register:
 *     that double-executes commands.
 *   * SCREENSHOTS TO DISK - a returned png is written to ../shots and replaced with a
 *     path, rather than pushing a huge base64 blob back through the tool result.
 */
const http = require('http');
const fs = require('fs');
const path = require('path');

const PORT = parseInt(process.argv[2] || '8765', 10);
const REDELIVER_AFTER = 35000; // ms before an un-acked command goes to someone else

const HERE = __dirname;
const PKG = path.dirname(HERE);
const SHOTS = path.join(PKG, 'shots');
const PRO_STATE = path.join(PKG, 'pro.state');

try { fs.mkdirSync(SHOTS, { recursive: true }); } catch (e) {}

let queue = [];             // commands waiting for the extension
const results = new Map();  // id -> result
const inFlight = new Map(); // id -> {cmd, ts}
let generation = 0;
let counter = 0;

function readPro() {
  // DEFAULT OFF. Google AI Pro is a paid plan and not every user has one - nobody gets
  // silently switched into a mode they aren't paying for, and nobody's results quietly
  // change without them asking. Whoever wants it turns it on once and it sticks.
  try {
    return fs.readFileSync(PRO_STATE, 'utf8').trim().toLowerCase() === 'on';
  } catch (e) {
    return false;
  }
}

function writePro(on) {
  try { fs.writeFileSync(PRO_STATE, on ? 'on' : 'off'); } catch (e) {}
}

function send(res, code, ctype, body) {
  const buf = Buffer.isBuffer(body) ? body : Buffer.from(String(body), 'utf8');
  res.writeHead(code, {
    'Content-Type': ctype,
    'Content-Length': buf.length,
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
    'Access-Control-Allow-Headers': 'content-type',
    'Connection': 'close'
  });
  res.end(buf);
}

function sendJson(res, code, obj) {
  send(res, code, 'application/json', JSON.stringify(obj));
}

function readBody(req) {
  return new Promise((resolve) => {
    let data = '';
    req.on('data', (c) => { data += c; if (data.length > 64 * 1024 * 1024) req.destroy(); });
    req.on('end', () => {
      try { resolve(data ? JSON.parse(data) : {}); } catch (e) { resolve({}); }
    });
    req.on('error', () => resolve({}));
  });
}

const server = http.createServer(async (req, res) => {
  const u = new URL(req.url, 'http://127.0.0.1');
  const p = u.pathname;
  const qp = u.searchParams;
  const method = req.method || 'GET';

  if (method === 'OPTIONS') return send(res, 204, 'text/plain', '');

  if (method === 'GET' && p === '/health') {
    return sendJson(res, 200, {
      ok: true, queued: queue.length, pending_results: results.size, port: PORT
    });
  }

  if (p === '/register_generation') {
    // A fresh background page announces itself: bump the generation so any older
    // zombie poll loop gets 410 and dies. Do NOT requeue in-flight work here - if the
    // old consumer finishes and acks it is done; if it died the reaper redelivers.
    generation += 1;
    return sendJson(res, 200, { generation });
  }

  if (method === 'GET' && p === '/poll') {
    const cg = parseInt(qp.get('generation') || '-1', 10);
    if (generation > 0 && (isNaN(cg) || cg < generation)) {
      return sendJson(res, 410, { error: 'ZOMBIE_REJECTED', generation });
    }
    const now = Date.now();
    for (const [id, v] of [...inFlight.entries()]) {
      if (now - v.ts > REDELIVER_AFTER) { inFlight.delete(id); queue.unshift(v.cmd); }
    }
    if (queue.length) {
      const cmd = queue.shift();
      inFlight.set(cmd.id, { cmd, ts: Date.now() });
      return sendJson(res, 200, cmd);
    }
    return send(res, 200, 'application/json', '{}');
  }

  if (p === '/ack') {
    const id = qp.get('id');
    if (id !== null) { inFlight.delete(parseInt(id, 10)); inFlight.delete(id); }
    return sendJson(res, 200, { ok: true });
  }

  if (method === 'POST' && p === '/command') {
    const data = await readBody(req);
    counter += 1;
    const id = counter;
    queue.push({ id, action: data.action || '', params: data.params || {} });
    return sendJson(res, 200, { id });
  }

  if (method === 'POST' && p === '/result') {
    const data = await readBody(req);
    if (data.id === undefined || data.id === null) {
      return sendJson(res, 400, { error: 'missing id' });
    }
    let result = data.result;
    if (result && typeof result === 'object' && result.png_base64) {
      // Write the screenshot out and hand back a path instead of a huge blob.
      try {
        const file = path.join(SHOTS, 'shot-' + data.id + '.png');
        fs.writeFileSync(file, Buffer.from(result.png_base64, 'base64'));
        delete result.png_base64;
        result.png_path = file;
      } catch (e) {
        delete result.png_base64;
        result.error = 'could not write screenshot: ' + e.message;
      }
    }
    results.set(parseInt(data.id, 10), result);
    return sendJson(res, 200, { ok: true });
  }

  const m = p.match(/^\/result\/(\d+)$/);
  if (method === 'GET' && m) {
    const id = parseInt(m[1], 10);
    if (results.has(id)) {
      const r = results.get(id);
      results.delete(id);
      return sendJson(res, 200, { done: true, result: r });
    }
    return sendJson(res, 200, { done: false });
  }

  if (method === 'GET' && p === '/pro') return sendJson(res, 200, { on: readPro() });

  if (method === 'POST' && p === '/pro') {
    const data = await readBody(req);
    writePro(!!data.on);
    return sendJson(res, 200, { ok: true, on: !!data.on });
  }

  if (method === 'GET' && (p === '/' || p === '/favicon.ico')) {
    return send(res, 200, 'text/plain', 'Claude Firefox Bridge running on port ' + PORT + '\n');
  }

  return sendJson(res, 404, { error: 'no such route', method, path: p });
});

server.listen(PORT, '127.0.0.1');
